File: be/src/io/fs/remote_file_system.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "io/fs/remote_file_system.h"

#include <bthread/bthread.h>
#include <glog/logging.h>

#include "common/config.h"
#include "io/fs/file_reader.h"

namespace doris::io {

Status RemoteFileSystem::upload(const Path& local_file, const Path& dest_file) {
    DCHECK(bthread_self() == 0) << "RemoteFileSystem::upload should not be called in bthread";
    Path dest_path;
    RETURN_IF_ERROR(absolute_path(dest_file, dest_path));
    return upload_impl(local_file, dest_path);
}

Status RemoteFileSystem::batch_upload(const std::vector<Path>& local_files,
                                      const std::vector<Path>& remote_files) {
    DCHECK(bthread_self() == 0) << "RemoteFileSystem::batch_upload should not be called in bthread";
    std::vector<Path> remote_paths;
    for (const auto& path : remote_files) {
        Path abs_path;
        RETURN_IF_ERROR(absolute_path(path, abs_path));
        remote_paths.push_back(abs_path);
    }
    return batch_upload_impl(local_files, remote_paths);
}

Status RemoteFileSystem::download(const Path& remote_file, const Path& local) {
    DCHECK(bthread_self() == 0) << "RemoteFileSystem::download should not be called in bthread";
    Path remote_path;
    RETURN_IF_ERROR(absolute_path(remote_file, remote_path));
    return download_impl(remote_path, local);
}

Status RemoteFileSystem::open_file_impl(const Path& path, FileReaderSPtr* reader,
                                        const FileReaderOptions* opts) {
    FileReaderSPtr raw_reader;
    if (!opts) {
        opts = &FileReaderOptions::DEFAULT;
    }
    RETURN_IF_ERROR(open_file_internal(path, &raw_reader, *opts));
    *reader = DORIS_TRY(create_cached_file_reader(raw_reader, *opts));
    return Status::OK();
}

} // namespace doris::io

File: fe/fe-core/src/main/java/org/apache/doris/fs/remote/RemoteFileSystem.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.fs.remote;

import org.apache.doris.analysis.StorageBackend;
import org.apache.doris.backup.Status;
import org.apache.doris.common.UserException;
import org.apache.doris.fs.PersistentFileSystem;

import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;

import java.io.Closeable;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

public abstract class RemoteFileSystem extends PersistentFileSystem implements Closeable {

    protected AtomicBoolean closed = new AtomicBoolean(false);

    public RemoteFileSystem(String name, StorageBackend.StorageType type) {
        super(name, type);
    }

    protected FileStatus[] getFileStatuses(String remotePath, FileSystem fileSystem) throws IOException {
        return fileSystem.listStatus(new Path(remotePath));
    }

    @Override
    public Status renameDir(String origFilePath,
                            String destFilePath,
                            Runnable runWhenPathNotExist) {
        Status status = exists(destFilePath);
        if (status.ok()) {
            return new Status(Status.ErrCode.COMMON_ERROR, "Destination directory already exists: " + destFilePath);
        }

        String targetParent = new Path(destFilePath).getParent().toString();
        status = exists(targetParent);
        if (Status.ErrCode.NOT_FOUND.equals(status.getErrCode())) {
            status = makeDir(targetParent);
        }
        if (!status.ok()) {
            return new Status(Status.ErrCode.COMMON_ERROR, status.getErrMsg());
        }

        runWhenPathNotExist.run();

        return rename(origFilePath, destFilePath);
    }

    public boolean connectivityTest(List<String> filePaths) throws UserException {
        return true;
    }
}
